src/Microsoft.AspNet.Authentication.OpenIdConnect/OpenIdConnectAuthenticationHandler.cs

```diff
@@ -520,6 +557,73 @@ protected override AuthenticationTicket AuthenticateCore()
             }
         }
 
+        protected virtual async Task<OpenIdConnectMessage> RedeemAuthorizationCode(string authorizationCode)
+        {
+            var openIdMessage = new OpenIdConnectMessage()
+            {
+                ClientId = Options.ClientId,
+                ClientSecret = Options.ClientSecret,
+                Code = authorizationCode,
+                GrantType = "authorization_code",
+                RedirectUri = Options.RedirectUri
+            };
+
+            var requestMessage = new HttpRequestMessage(HttpMethod.Post, _configuration.TokenEndpoint);
+            requestMessage.Content = new FormUrlEncodedContent(openIdMessage.Parameters);
+            var responseMessage = await Backchannel.SendAsync(requestMessage);
+            responseMessage.EnsureSuccessStatusCode();
+            var tokenResonse = await responseMessage.Content.ReadAsStringAsync();
+            var jsonTokenResponse = JObject.Parse(tokenResonse);
+            return new OpenIdConnectMessage()
+            {
+                AccessToken = jsonTokenResponse.Value<string>("access_token"),
+                IdToken = jsonTokenResponse.Value<string>("id_token"),
+                TokenType = jsonTokenResponse.Value<string>("token_type"),
+                ExpiresIn = jsonTokenResponse.Value<string>("expires_in")
+            };
+        }
+
+        protected virtual async Task<AuthenticationTicket> GetUserInformationAsync(AuthenticationProperties properties, OpenIdConnectMessage message, AuthenticationTicket ticket)
+        {
+            var requestMessage = new HttpRequestMessage(HttpMethod.Get, _configuration.UserInfoEndpoint);
```

Comment:
According to the OIDC discovery specifications, `userinfo_endpoint` is not a "required" parameter but a "recommended" one. Chances are high to crash here if it was missing from the discovery response and if the developer has not provided one in `app.UseOpenIdConnectAuthentication`.

Any chance you could update `GetUserInformationAsync` to work in a degraded mode in this case?

http://openid.net/specs/openid-connect-discovery-1_0.html#ProviderMetadata